### PR TITLE
Fix flaky SRIOV tests suite: Perform teardown before each test start

### PR DESCRIPTION
**Pull request by ormergi**
Signed-off-by: Or Mergi <ormergi@redhat.com>

**What this PR does / why we need it**:
The SRIOV test suite designed to avoid using the the general func for test teardown [1] before each test starts.
Instead [1] is called once before all tests starts, due to that VMI's clean up is done in-place before each test starts.

The VMI's clean up doesn't handle VMI's that take longer time to dispose or stuck in the Terminating phase.
Due to the fact that VMI's clean up occurs before tests starts, when it fails the test that about to start reported as it was failed instead of the one that just finished.

This PR Changes SRIOV test suite to use [1] before each test starts, it will clean up VMI's properly and prevent reporting wrong test failures.

**Which issue(s) this PR fixes** :

**Special notes for your reviewer**:
In order to make use of [1], NAD's creating is moved to tests spec instead before all tests start.

Here we can see that the test `[test_id:3985]should create a virtual machine with sriov interface with custom MAC` reported as failed [2] [3], but did not even started.
Looking into the VMI state from artifacts [4] show its the VMI spec from the test before
[5] and not the one that reported as failure.

[1] https://github.com/kubevirt/kubevirt/blob/8f8a93245fc6bbaf881470f0ec97dcdae5ede8b8/tests/utils.go#L491
[2] https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6231/pull-kubevirt-e2e-kind-1.19-sriov/1425816945118154752
[3] https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6182/pull-kubevirt-e2e-kind-1.19-sriov/1425895209576697856
[4] https://storage.googleapis.com/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6182/pull-kubevirt-e2e-kind-1.19-sriov/1425895209576697856/artifacts/k8s-reporter/1_vmis.log
[5] https://github.com/kubevirt/kubevirt/blob/8f8a93245fc6bbaf881470f0ec97dcdae5ede8b8/tests/network/vmi_multus.go#L825-L832

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: tests/network/vmi_multus.go [hunk 1]
```diff
@@ -761,6 +748,8 @@ var _ = Describe("[Serial]SRIOV", func() {
 		}
 
 		It("should block migration for SR-IOV VMI's when LiveMigration feature-gate is on but SRIOVLiveMigration is off", func() {
+			Expect(createNetworkAttachementDefinition(sriovnet1, util.NamespaceTestDefault, sriovConfCRD)).To((Succeed()), "should successfully create the network")
```
Comment (EdDev) [inline tests/network/vmi_multus.go:751]: I prefer for this change, just to place it all in the `BeforeEach` and not distribute it all over.
We need to re-organize the tests here, lets defer this part to that step.

**Review by EdDev — CHANGES_REQUESTED** [review]
Looks like a good opertunity to remove the `BeforeAll`.
See details inline.

**Review by EdDev — APPROVED** [review]
Lets start with something and then we will worry about the rest.

Thanks!

**Comment by ormergi** [comment]
> @ormergi unit tests [failing](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6267/pull-kubevirt-unit-test/1431065531342917632)?!
> 
> ```

I think its transient or has been fixed, this PR doesn't even touch the code that could affect those tests

**Comment by ormergi** [comment]
> @ormergi unit tests [failing](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6267/pull-kubevirt-unit-test/1431065531342917632)?!
> 
> ```
>  E0827 01:35:05.946225      12 runtime.go:76] Observed a panic: sync: negative WaitGroup counter 
> ```
> 
@dankenigsberg I think its transient or has been fixed, this PR doesn't even touch code that could affect those tests

**Comment by dankenigsberg** [comment]
@ormergi clearly this PR is not causing these failed unit test; but we ensure that unit test cannot be flaky. I see this two often.

**Comment by ormergi** [comment]
@davidvossel @enp0s3 I think we have a problem with k8s-1.20-sig-storage lane, since this PR wont affect storage tests can we override it?

**Comment by dankenigsberg** [comment]
> @davidvossel @enp0s3 I think we have a problem with k8s-1.20-sig-storage lane, since this PR wont affect storage tests can we override it?

The storage sig, say @maya-r @awels @aglitke, should be aware of this.

**Comment by maya-r** [comment]
I've proposed https://github.com/kubevirt/project-infra/pull/1121 to avoid blocking kubevirt asap (while looking for more fixes)

**Comment by ormergi** [comment]
> I've proposed [kubevirt/project-infra#1121](https://github.com/kubevirt/project-infra/pull/1121) to avoid blocking kubevirt asap (while looking for more fixes)

I am not sure how its going to mitigate k8s-1.20-sig-storage lane, here we can see it keeps blocking

**Comment by maya-r** [comment]
oops. https://github.com/kubevirt/project-infra/pull/1520

**Comment by ormergi** [comment]
> oops. [kubevirt/project-infra#1520](https://github.com/kubevirt/project-infra/pull/1520)

Thanks!

@fedepaol I see that merging kubevirt/project-infra#1520 caused retest for all lanes, is it correct?
1.20 and 1.19 sig-storage  lanes are still marked as required, would it be merged anyway?

**Comment by fgimenez** [comment]
> @fedepaol I see that merging [kubevirt/project-infra#1520](https://github.com/kubevirt/project-infra/pull/1520) caused retest for all lanes, is it correct?

I see that lanes that already passed are not being rebuilt right? like https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6267/pull-kubevirt-e2e-k8s-1.19-sig-storage/1431804818728423424, is that correct? and poor Federico Paolinelli :)  

> 1.20 and 1.19 sig-storage lanes are still marked as required, would it be merged anyway?

where do you see that? In the summary of the checks I see that 1.20-sig-storage and 1.19-sig-storage don't have the "Required" label anymore, see screenshot attached.

Screenshot from 2021-08-30 11-02-45

**Comment by fedepaol** [comment]
> > @fedepaol I see that merging [kubevirt/project-infra#1520](https://github.com/kubevirt/project-infra/pull/1520) caused retest for all lanes, is it correct?
> 
> I see that lanes that already passed are not being rebuilt right? like https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6267/pull-kubevirt-e2e-k8s-1.19-sig-storage/1431804818728423424, is that correct? and poor Federico Paolinelli :)
> 


Lol I don't mind remembering my old project from time to time :P


> > 1.20 and 1.19 sig-storage lanes are still marked as required, would it be merged anyway?
> 
> where do you see that? In the summary of the checks I see that 1.20-sig-storage and 1.19-sig-storage don't have the "Required" label anymore, see screenshot attached.
> 
> Screenshot from 2021-08-30 11-02-45

**Comment by ormergi** [comment]
> I see that lanes that already passed are not being rebuilt right? like https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6267/pull-kubevirt-e2e-k8s-1.19-sig-storage/1431804818728423424, is that correct? and poor Federico Paolinelli :)
> 

Ohh I am so sorry my bad :sweat:

I am not sure because I see that the rest of sig lanes are triggered

> where do you see that? In the summary of the checks I see that 1.20-sig-storage and 1.19-sig-storage don't have the "Required" label anymore, see screenshot attached.
> 

I guess its updated now, when writing that comment it was marked as required

**Comment by fgimenez** [comment]
> I am not sure because I see that the rest of sig lanes are triggered

Yes, maybe related to how tide works with merge batches and not related to the change in the storage lanes, in any case it seems that already green lanes are not re-built.

**Comment by fgimenez** [comment]
`pull-kubevirt-e2e-k8s-1.21-sig-network` failed with a test that was removed from quarantine recently https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6267/pull-kubevirt-e2e-k8s-1.21-sig-network/1432264444922040320

**Comment by EdDev** [comment]
> `pull-kubevirt-e2e-k8s-1.21-sig-network` failed with a test that was removed from quarantine recently https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6267/pull-kubevirt-e2e-k8s-1.21-sig-network/1432264444922040320

This seems a different issue than the one originally quarantined and fixed.
The IP address for the secondary interface is not reflected on the VMI state [1], although already set in the guest.

We need to monitor this one, to see if it occurs frequent. The 2 minutes seems to be enough, so it is strange to see this occurring.

[1]
```
                "interfaces": [
                    {
                        "ipAddress": "10.244.196.159",
                        "mac": "52:54:00:d0:2a:14",
                        "name": "default",
                        "ipAddresses": [
                            "10.244.196.159",
                            "fd10:244::c49d"
                        ],
                        "interfaceName": "eth0"
                    },
                    {
                        "mac": "02:03:04:05:06:aa",
                        "name": "net1",
                        "interfaceName": "eth1"
                    }
                ]
```

**Comment by ormergi** [comment]
> > `pull-kubevirt-e2e-k8s-1.21-sig-network` failed with a test that was removed from quarantine recently https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6267/pull-kubevirt-e2e-k8s-1.21-sig-network/1432264444922040320
> 
> This seems a different issue than the one originally quarantined and fixed.
> The IP address for the secondary interface is not reflected on the VMI state [1], although already set in the guest.
> 
> We need to monitor this one, to see if it occurs frequent. The 2 minutes seems to be enough, so it is strange to see this occurring.
> 
> [1]
> 
> ```
>                 "interfaces": [
>                     {
>                         "ipAddress": "10.244.196.159",
>                         "mac": "52:54:00:d0:2a:14",
>                         "name": "default",
>                         "ipAddresses": [
>                             "10.244.196.159",
>                             "fd10:244::c49d"
>                         ],
>                         "interfaceName": "eth0"
>                     },
>                     {
>                         "mac": "02:03:04:05:06:aa",
>                         "name": "net1",
>                         "interfaceName": "eth1"
>                     }
>                 ]
> ```

I think it happens because we assert on the VMI status w/o making sure QEMU guest agent is up.
I posted a PR to add it https://github.com/kubevirt/kubevirt/pull/6318

**Comment by fgimenez** [comment]
> I think it happens because we assert on the VMI status w/o making sure QEMU guest agent is up.
> I posted a PR to add it #6318
> 
> EDIT: I am not sure its the root cause, but its still necessary.

thanks a lot, another failure here https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6312/pull-kubevirt-e2e-k8s-1.21-sig-network/1432303029054345216

**Pull request merged**